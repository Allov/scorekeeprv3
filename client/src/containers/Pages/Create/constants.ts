export const CREATE_GAME = 'CREATE_GAME';
export type CREATE_GAME = typeof CREATE_GAME;

export const INPUT_CHANGED = 'INPUT_CHANGED';
export type INPUT_CHANGED = typeof INPUT_CHANGED;

export const SET_GAME_NAME = 'SET_GAME_NAME';
export type SET_GAME_NAME = typeof SET_GAME_NAME;

export const SET_GAME_NUMBEROFPLAYERS = 'SET_GAME_NUMBEROFPLAYERS';
export type SET_GAME_NUMBEROFPLAYERS = typeof SET_GAME_NUMBEROFPLAYERS;

export const SET_GAME_USERID = 'SET_GAME_USERID';
export type SET_GAME_USERID = typeof SET_GAME_USERID;
