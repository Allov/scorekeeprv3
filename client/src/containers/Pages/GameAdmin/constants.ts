export const FETCH_GAME = 'FETCH_GAME';
export type FETCH_GAME = typeof FETCH_GAME;

export const FETCHED_GAME = 'FETCHED_GAME';
export type FETCHED_GAME = typeof FETCHED_GAME;

export const ADDPLAYERTO_GAME = 'ADDPLAYERTO_GAME';
export type ADDPLAYERTO_GAME = typeof ADDPLAYERTO_GAME;

export const ADDEDPLAYERTO_GAME = 'ADDEDPLAYERTO_GAME';
export type ADDEDPLAYERTO_GAME = typeof ADDEDPLAYERTO_GAME;
