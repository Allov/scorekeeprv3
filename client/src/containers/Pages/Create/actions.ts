import { Action } from 'redux';
import * as constants from './constants';
import { IGameInput } from '../../../types';

export interface ICreateGameAction extends Action {
  type: constants.CREATE_GAME;
}

export function createGame() : ICreateGameAction {
  return {
    type: constants.CREATE_GAME,
  };
}

export interface IInputChanged extends Action {
  input: IGameInput;
}

export function inputChanged(input: IGameInput) : IInputChanged {
  return {
    input,
    type: constants.INPUT_CHANGED,
  };
}

export interface ISetGameName extends Action {
  name: string;
}

export function setGameName(name: string) : ISetGameName {
  return {
    name,
    type: constants.SET_GAME_NAME,
  };
}

export interface ISetGameNumberOfPlayers extends Action {
  numberOfPlayers: number;
}

export function setGameNumberOfPlayers(numberOfPlayers: number) : ISetGameNumberOfPlayers {
  return {
    numberOfPlayers,
    type: constants.SET_GAME_NUMBEROFPLAYERS,
  };
}

export type CreateGameActions = ICreateGameAction | IInputChanged | ISetGameName | ISetGameNumberOfPlayers | undefined;
