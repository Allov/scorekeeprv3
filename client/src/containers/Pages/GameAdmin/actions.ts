import { Action } from 'redux';
import { IGame, IIndexedAction } from '../../../types';
import * as constants from './constants';

export interface IFetchGameAction extends Action {
  shareId: string;
  type: constants.FETCH_GAME;
}

export function fetchGame(shareId: string): IFetchGameAction {
  return {
    shareId,
    type: constants.FETCH_GAME,
  };
}

export interface IFetchedGameAction extends Action {
  game: IGame;
  type: constants.FETCHED_GAME;
}

export function fetchedGame(game: IGame): IFetchedGameAction {
  return {
    game,
    type: constants.FETCHED_GAME,
  };
}

export interface IAddPlayerToGameAction extends Action {
  gameId: string;
  name: string;
}

export function addPlayerToGame(gameId: string, name: string): IAddPlayerToGameAction {
  return {
    gameId,
    name,
    type: constants.ADDPLAYERTO_GAME,
  };
}

export interface IAddedPlayerToGameAction extends Action {
  game: IGame;
}

export function addedPlayerToGame(game: IGame): IAddedPlayerToGameAction {
  return {
    game,
    type: constants.ADDEDPLAYERTO_GAME,
  };
}

export interface IEditedPlayerNameAction extends Action, IIndexedAction {
  name: string;
}

export function editedPlayerName(index: number, name: string): IEditedPlayerNameAction {
  return {
    index,
    name,
    type: constants.EDITEDPLAYER_NAME,
  };
}

export interface IEditedPlayerPointsAction extends Action, IIndexedAction {
  points: string;
}

export function editedPlayerPoints(index: number, points: string): IEditedPlayerPointsAction {
  return {
    index,
    points,
    type: constants.EDITEDPLAYER_POINTS,
  };
}

export interface IDeletePlayerFromGame extends Action {
  id: string,
}

export function deletePlayerFromGame(id: string): IDeletePlayerFromGame {
  return {
    id,
    type: constants.DELETEPLAYERFROM_GAME,
  };
}

export type GameAdminActions =
  IFetchGameAction |
  IFetchedGameAction |
  IAddPlayerToGameAction |
  IAddedPlayerToGameAction |
  IEditedPlayerNameAction |
  IEditedPlayerPointsAction |
  IDeletePlayerFromGame |
  undefined;
