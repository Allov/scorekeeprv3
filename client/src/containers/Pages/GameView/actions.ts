import { Action } from 'redux';
import { IGame } from 'src/types';
import * as constants from './constants';

export interface ISubscribeGameAction extends Action {
  shareId: string;
  type: constants.SUBSCRIBETO_GAME;
}

export function subscribeToGame(shareId: string): ISubscribeGameAction {
  return {
    shareId,
    type: constants.SUBSCRIBETO_GAME,
  };
}

export interface ISubscribedToGameAction extends Action {
  game: IGame;
  type: constants.SUBSCRIBEDTO_GAME;
}

export function subscribedToGame(game: IGame): ISubscribedToGameAction {
  return {
    game,
    type: constants.SUBSCRIBEDTO_GAME,
  };
}

export interface ISubscribedGameUpdatedAction extends Action {
  game: IGame;
  type: constants.SUBSCRIBEDGAME_UPDATE;
}

export function subscribedGameUpdated(game: IGame): ISubscribedGameUpdatedAction {
  return {
    game,
    type: constants.SUBSCRIBEDGAME_UPDATE,
  };
}
