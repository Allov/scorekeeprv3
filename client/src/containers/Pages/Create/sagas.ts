import { gql } from 'apollo-boost';
import { push } from 'connected-react-router';
import { call, put, takeLatest } from 'redux-saga/effects';
import client from '../../../lib/apollo';
import { Notifications } from '../../../types/constants';
import { notify } from '../../Notifications/actions';
import { CREATE_GAME } from './constants';

const QUERY = gql`
  mutation {
    addGame(input: {
      name: "anonymous game",
      userId: null,
    }) {
      shareId
    }
  }
`;

export function* createGame() {

  try {
    const result = yield call(client.mutate, { mutation: QUERY });
    // todo: type result.
    yield put(push(`/game/${result.data.addGame.shareId}`));
  } catch(error) {
    // tslint:disable-next-line:no-console
    console.error(error);

    yield put(notify(Notifications.Error, `Could not create game: ${error.message}`));
  }
}

export const createSagas = [
  takeLatest(CREATE_GAME, createGame),
];
