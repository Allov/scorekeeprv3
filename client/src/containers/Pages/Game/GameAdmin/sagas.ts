import { gql } from 'apollo-boost';
import { put, select, takeLatest, throttle } from 'redux-saga/effects';
import { IRound } from '../../../../types';
import { safeMutate, safeQuery } from '../../../../lib/saga';
import { Notifications } from '../../../../types/constants';
import { notify } from '../../../Notifications/actions';
import { fetchedGame, IAddPlayerToGameAction, IDeletePlayerFromGame, IEditedPlayerNameAction, IEditedPlayerPointsAction, IFetchGameAction } from '../actions';
import { ADDPLAYERTO_GAME, DELETEPLAYERFROM_GAME, EDITEDPLAYER_NAME, EDITEDPLAYER_POINTS, FETCH_GAME } from '../constants';
import { makeSelectGameCurrentRound, makeSelectGameId, makeSelectPlayerId } from '../selectors';

const GAME_FRAGMENT = gql`
  fragment GameFields on Game {
    id,
    name,
    shareId,
    players {
      id,
      name,
    }
    rounds {
      id,
      roundNumber,
      scores {
        points,
        player {
          id,
          name,
          totalScore,
        }
      }
    },
    createdBy {
      username
    }
  }
`;

const FETCH_GAME_BY_SHAREID = gql`
  query GameByShareId($shareId: String!) {
    gameByShareId(shareId: $shareId) {
      ...GameFields,
    }
  }

  ${GAME_FRAGMENT}
`;

export function* fetchGame(action: IFetchGameAction) {
  const result = yield safeQuery(FETCH_GAME_BY_SHAREID, {
    shareId: action.shareId,
  },
    'Could not fetch game');

  if (!result.gameByShareId) {
    yield put(notify(Notifications.Error, `Could not find game: ${action.shareId}`));
  } else {
    yield put(fetchedGame(result.gameByShareId));
  }
}

const ADD_PLAYER = gql`
  mutation AddPlayerToGame($input: PlayerInput!) {
    addPlayerToGame(input: $input) {
      ...GameFields,
    }
  }

  ${GAME_FRAGMENT}
`;

export function* addPlayerToGame(action: IAddPlayerToGameAction) {
  const result = yield safeMutate(ADD_PLAYER, {
    input: {
      gameId: action.gameId,
      name: action.name,
    },
  },
    'Could not add player to game'
  );

  // since playerId is generated on the server side, let's synchronise the game.
  // new player should be already added in the array with the same name unless it was so fast and began typing a name.
  yield put(fetchedGame(result.addPlayerToGame));
}

const UPDATE_PLAYER = gql`
  mutation updatePlayer($playerId: String!, $input: PlayerInput!) {
    updatePlayer(id: $playerId, input: $input) {
      id,
    }
  }
`;

export function* editedPlayerName(action: IEditedPlayerNameAction) {
  const gameId = yield select(makeSelectGameId());
  const playerId = action.id;

  yield safeMutate(UPDATE_PLAYER, {
    input: {
      gameId,
      name: action.name,
    },
    playerId,
  },
    'Could not edit player name'
  );
}

const UPDATE_SCORES = gql`
  mutation updateScores($input: ScoresInput!) {
    updateScores(input: $input) {
      id,
    }
  }
`;

export function* editedPlayerPoints(action: IEditedPlayerPointsAction) {
  const gameId = yield select(makeSelectGameId());
  const round: IRound = yield select(makeSelectGameCurrentRound());
  const playerId = action.id;

  yield safeMutate(UPDATE_SCORES, {
    input: {
      filter: {
        gameId,
        roundId: round.id,
      },
      scores: [{
        playerId,
        points: Number(action.points),
      }],
    },
  },
    'Could not update player score'
  );
}

const DELETE_PLAYER = gql`
  mutation deletePlayerFromGame($playerId: String!, $gameId: String!) {
    deletePlayer(id: $playerId, gameId: $gameId) {
      id
    }
  }
`;

export function* deletedPlayerFromGame(action: IDeletePlayerFromGame) {
  const gameId = yield select(makeSelectGameId());

  yield safeMutate(DELETE_PLAYER, {
    gameId,
    playerId: action.id,
  },
    'Could not delete player'
  );
}

export const gameAdminSagas = [
  takeLatest(FETCH_GAME, fetchGame),
  takeLatest(ADDPLAYERTO_GAME, addPlayerToGame),
  throttle(500, EDITEDPLAYER_NAME, editedPlayerName),
  takeLatest(EDITEDPLAYER_POINTS, editedPlayerPoints),
  takeLatest(DELETEPLAYERFROM_GAME, deletedPlayerFromGame),
];
