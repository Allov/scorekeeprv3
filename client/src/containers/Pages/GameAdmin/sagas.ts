import { gql } from 'apollo-boost';
import { call, put, takeLatest } from 'redux-saga/effects';
import client from '../../../lib/apollo';
import { Notifications } from '../../../types/constants';
import { notify } from '../../Notifications/actions';
import { fetchedGame, IFetchGameAction } from './actions';
import { FETCH_GAME } from './constants';

const QUERY = gql`
  query GameByShareId($shareId: String!) {
    gameByShareId(shareId: $shareId) {
      id,
      name,
      shareId,
      createdBy {
        username
      }
    }
  }
`;

export function* fetchGame(action: IFetchGameAction) {
  try {
    const result = yield call(client.query, { query: QUERY, variables: { shareId: action.shareId } });

    if (!result.data.gameByShareId) {
      yield put(notify(Notifications.Error, `Could not find game: ${action.shareId}`));
    } else {
      yield put(fetchedGame(result.data.gameByShareId));
    }
  } catch(error) {
    // tslint:disable-next-line:no-console
    console.error(error);

    yield put(notify(Notifications.Error, `Could not fetch game: ${error.message}`));
  }
}

export const gameAdminSagas = [
  takeLatest(FETCH_GAME, fetchGame),
];
