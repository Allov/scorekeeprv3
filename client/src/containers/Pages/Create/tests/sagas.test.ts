import { push } from 'connected-react-router';
import { put } from 'redux-saga/effects';
import { Notifications } from '../../../../types/constants';
import { notify } from '../../../Notifications/actions';
import { createGame } from '../sagas';

let createGameGenerator = createGame();

beforeEach(() => {
  createGameGenerator = createGame();
  const callDescriptor = createGameGenerator.next().value;
  expect(callDescriptor).toMatchSnapshot();
})

it('dispatch the push action if game was created correctly', () => {
  const shareId = 'dont-care';
  const result = {
    data: {
      createGame: {
        shareId,
      }
    }
  };

  const putDescriptor = createGameGenerator.next(result).value;
  expect(putDescriptor).toEqual(put(push(`/game/${shareId}/admin`)));
});

it('dispatch a notification error on error', () => {
  const response = new Error('Woops');
  // compiler gives 'cannot invoke an object which is possibly 'undefined'' and I can't find anything to fix this to save my life.
  // casting createGameGenerator as any here fixes the compiler error and the code runs just fine.
  const putDescriptor = (createGameGenerator as any).throw(response).value;

  expect(putDescriptor).toEqual(put(notify(Notifications.Error, `Could not create game: ${response.message}`)));
});
