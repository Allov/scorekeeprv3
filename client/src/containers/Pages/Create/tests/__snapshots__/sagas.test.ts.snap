// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatch a notification error on error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "mutation": Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": undefined,
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "input",
                        },
                        "value": Object {
                          "fields": Array [
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "value": Object {
                                "block": false,
                                "kind": "StringValue",
                                "value": "anonymous game",
                              },
                            },
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "userId",
                              },
                              "value": Object {
                                "kind": "NullValue",
                              },
                            },
                          ],
                          "kind": "ObjectValue",
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "createGame",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "shareId",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 121,
            "start": 0,
          },
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`dispatch the push action if game was created correctly 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "mutation": Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": undefined,
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": undefined,
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "input",
                        },
                        "value": Object {
                          "fields": Array [
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "value": Object {
                                "block": false,
                                "kind": "StringValue",
                                "value": "anonymous game",
                              },
                            },
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "userId",
                              },
                              "value": Object {
                                "kind": "NullValue",
                              },
                            },
                          ],
                          "kind": "ObjectValue",
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "createGame",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "shareId",
                          },
                          "selectionSet": undefined,
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 121,
            "start": 0,
          },
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
